Rn remove ccs record tableau

Rohit25negi opened the pull request:
This code removes the dependency on ccs_monthy_tableau from ccs_record_monthly.
A few new columns are added to ccs_record_cases.json, birth_preparedness_forms and delivery form for achieve this.
I am not removing ccs tableau because to enable easy revert in case of problem and because it is also used in `custom/icds_reports/utils/aggregation_helpers/agg_ccs_record.py`. Will be creating a separate PR for that.

Please go through the each commit that will be easier to review. 
I added the comments on each comment. Please review by commit for easier review
Please do not merge until these are deployed: https://github.com/dimagi/commcare-hq/pull/22973, https://github.com/dimagi/commcare-hq/pull/23109

Rohit25negi committed:
REmoving in progress

Rohit25negi committed:
WIP

Rohit25negi committed:
Replace ccs record tabealu columns with ccs record case

calellowitz commented:
Going to do a fuller review on individual indicators later this week, but two overall thoughts.
1) Why does all the test data change in the outputs? None of the indicators or data in the dashboard should change with this. Furthermore, it makes it really hard to tell if the indicators are implemented correctly because the underlying tests of correctness have changed.
2) A number of indicators (I remember noticing it with pregnant and lactating, but I know there were a few others) use only the current state, so if we reaggregate a past month, which we do every day, and the state has changed, the data will be incorrect in the past month. For example, if the woman gave birth in the current month, she will not be marked pregnant for either the current month or the past month, and if we have to manually aggregate a previous month, she will no longer be pregant in that month either.

Rohit25negi commented:
@calellowitz 
1. I needed to change the test data in the output because before case_monthly datasource was containing the field like age_in_month, trimester etc. which are calculated field and now sine test fixture needs to be added because it was not there. I added  `ccs_cases.csv` which contain the raw field like `add` , `edd` etc which are used to calculate the field mentioned above. It was becoming difficult to reverse calculate the in x `y=f(x)`.
2. I understood  what you are saying. I think that will be handled for lactating but not for pregnant as it is directly taken from the datasource. Lets offline about this

Rohit25negi committed:
Corrected the query

Rohit25negi committed:
Merge branch 'master' into rn_remove_ccs_record_tableau

Rohit25negi committed:
Merge branch 'master' into rn_remove_ccs_record_tableau

Rohit25negi committed:
After merged with master

Rohit25negi committed:
Make Birthpreparedness full join with prev month

Rohit25negi committed:
Corrected migration

Rohit25negi committed:
Added
date for dob,add,edd,opened_on

Rohit25negi committed:
Correct contradictory data of
resting_during_pregnany and counsel_immediate_bf

Rohit25negi committed:
Corrected date(only format)
Corected the consistency of extra_meal

Rohit25negi committed:
Corect conflicting data with pregnancy
1. if not pregnant, cannot be resting_in_pregnancy
2. if not pregnant, cannot be extra_meal

Rohit25negi committed:
Age in month can only increase by 1 month

Rohit25negi committed:
Putting anemia value from bp form

Rohit25negi committed:
Corrected test data for immediate_breasfeeding
adding data for test state in fixture

Rohit25negi committed:
Fix 'add' date

Rohit25negi committed:
set Bp visit in month=1 where bp_date lies in the month

Rohit25negi committed:
Set open_in_month=1 where pregnant=1

Rohit25negi committed:
Add bp_date with home_visit_date
they contains same info, not sure why its created

Rohit25negi committed:
corrected add date
NOTE: no pregnant and lactating output data touched
fixed all ccs record test

Rohit25negi committed:
added Dates same as ccs_record test output data
date for add, dob, bp_date,edd,opened_on

Rohit25negi committed:
Rolling up the conflict resolution to st1 test data
conflict were in:
anemia value
extra meal
eating_extra
immediate_breastfeeding
resting_during_pregnancy

Rohit25negi committed:
Final corrected script

Rohit25negi committed:
Added person case data
Add new fixtures to mapping
aggregation bp form for month 3

Rohit25negi committed:
Add bp and delivery form fixture

Rohit25negi commented:
@calellowitz  this is ready for review Added Both of your concerns

Rohit25negi committed:
Satisfy Lynter

Rohit25negi commented on custom/icds_reports/utils/aggregation_helpers/birth_preparedness_forms.py:
```diff
@@ -78,36 +80,38 @@ def aggregation_query(self):
+            COALESCE(ucr.using_ifa, prev_month.using_ifa) as using_ifa
           FROM ({ucr_table_query}) ucr
-          LEFT JOIN "{previous_month_tablename}" prev_month
+          FULL JOIN "{previous_month_tablename}" prev_month
```
To make it satisfy the need of ever submitted/last submitted form. (Eg. if pregnant check if bp form submitted with immediate_breastfeeding='yes' prior to the end of month)

reference:  https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/utils/aggregation_helpers/postnatal_care_forms_ccs_record.py#L69

Rohit25negi commented on custom/icds_reports/tests/outputs/ccs_record_monthly_2017-04-01_sorted.csv:
```diff
@@ -21,9 +21,9 @@ a1,b69e4bd0-5362-4441-9389-855422bb9dd1,2017-04-01,268,other,0,1,,0,0,0,0,0,0,0,
 a1,c498b719-cad3-4be2-b342-a9bd35128714,2017-04-01,304,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1992-01-10,0,9e8de64b-b6e6-482e-a71d-8d266d55e7bd
 a1,cc2e1551-94d4-41dc-a7be-cb1c78c4f42b,2017-04-01,268,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1995-01-08,0,a361753c-9ee5-408b-ae6b-ab807424a81d
-a1,d499215f-5e4e-4c97-8ea5-78c3a9984881,2017-04-01,304,lactating,1,1,,20,1,0,0,0,0,0,0,,0,0,1,0,0,0,1,1,0,,0,0,0,0,0,0,0,0,0,0,1,0,0,1,,0,0,1,1,0,2017-03-01,0,obc,no,no,yes,,,,,,,,,,,1,,1,,,2017-08-08,,,,,,,,,,1,,,,,,,2017-04-12,,1992-01-10,0,dde249e4-a7b1-4d2b-bee3-d0e93a8566bf
+a1,d499215f-5e4e-4c97-8ea5-78c3a9984881,2017-04-01,304,lactating,1,1,,20,1,0,0,0,0,0,0,,0,0,1,0,0,0,1,1,0,,0,0,0,0,0,0,0,0,0,0,1,0,0,1,,0,0,1,1,0,2017-03-01,0,obc,no,no,yes,,,,,,,,,,,1,1,1,,,2017-08-08,,,,,,,,,,1,,,,,,,2017-04-12,,1992-01-10,0,dde249e4-a7b1-4d2b-bee3-d0e93a8566bf
```
person whose extra_meal is 1 its eating_extra has to be 1

Rohit25negi commented on custom/icds_reports/tests/outputs/ccs_record_monthly_2017-04-01_sorted.csv:
```diff
@@ -21,9 +21,9 @@ a1,b69e4bd0-5362-4441-9389-855422bb9dd1,2017-04-01,268,other,0,1,,0,0,0,0,0,0,0,
 a1,c498b719-cad3-4be2-b342-a9bd35128714,2017-04-01,304,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1992-01-10,0,9e8de64b-b6e6-482e-a71d-8d266d55e7bd
 a1,cc2e1551-94d4-41dc-a7be-cb1c78c4f42b,2017-04-01,268,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1995-01-08,0,a361753c-9ee5-408b-ae6b-ab807424a81d
-a1,d499215f-5e4e-4c97-8ea5-78c3a9984881,2017-04-01,304,lactating,1,1,,20,1,0,0,0,0,0,0,,0,0,1,0,0,0,1,1,0,,0,0,0,0,0,0,0,0,0,0,1,0,0,1,,0,0,1,1,0,2017-03-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,1,,,,,,,2017-04-12,,1992-01-10,0,dde249e4-a7b1-4d2b-bee3-d0e93a8566bf
+a1,d499215f-5e4e-4c97-8ea5-78c3a9984881,2017-04-01,304,lactating,1,1,,20,1,0,0,0,0,0,0,,0,0,1,0,0,0,1,1,0,,0,0,0,0,0,0,0,0,0,0,1,0,0,1,,0,0,1,1,0,2017-03-01,0,obc,no,no,yes,,,,,,,,,,,1,,1,,,2017-08-08,,,,,,,,,,1,,,,,,,2017-04-12,,1992-01-10,0,dde249e4-a7b1-4d2b-bee3-d0e93a8566bf
```
if resting_during_pregnany=1 then resting must be 1
if counsel_immediate_bf=1 then immediate_breastfeeding has to be 1

Rohit25negi commented on custom/icds_reports/tests/outputs/ccs_record_monthly_2017-04-01_sorted.csv:
```diff
@@ -21,7 +21,7 @@ a1,b69e4bd0-5362-4441-9389-855422bb9dd1,2017-04-01,268,other,0,1,,0,0,0,0,0,0,0,
 a1,c498b719-cad3-4be2-b342-a9bd35128714,2017-04-01,304,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1992-01-10,0,9e8de64b-b6e6-482e-a71d-8d266d55e7bd
 a1,cc2e1551-94d4-41dc-a7be-cb1c78c4f42b,2017-04-01,268,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1995-01-08,0,a361753c-9ee5-408b-ae6b-ab807424a81d
-a1,d499215f-5e4e-4c97-8ea5-78c3a9984881,2017-04-01,304,lactating,1,1,,20,1,0,0,0,0,0,0,,0,0,1,0,0,0,1,1,0,,0,0,0,0,0,0,0,0,0,0,1,0,0,1,,0,0,1,1,0,2017-03-01,0,obc,no,no,yes,,,,,,,,,,,1,1,1,,,2017-08-08,,,,,,,,,,1,,,,,,,2017-04-12,,1992-01-10,0,dde249e4-a7b1-4d2b-bee3-d0e93a8566bf
+a1,d499215f-5e4e-4c97-8ea5-78c3a9984881,2017-04-01,304,lactating,1,1,,20,1,0,0,0,0,0,0,,0,0,1,0,0,0,0,0,0,,0,0,0,0,0,0,0,0,0,0,1,0,0,1,,0,0,1,1,0,2017-03-01,0,obc,no,no,yes,,,,,,,,,,,1,1,1,,,2017-08-08,,,,,,,,,,1,,,,,,,2017-04-12,,1992-01-10,0,dde249e4-a7b1-4d2b-bee3-d0e93a8566bf
```
could have changed the pregnant instead of resting_in_pregnancy and extra_meal But pregnant parameter is used in other places in icds also like pregnant and lacting women report. Did not want to propagate the change

Rohit25negi commented on custom/icds_reports/tests/outputs/ccs_record_monthly_2017-05-01_sorted.csv:
```diff
@@ -81,22 +81,22 @@ a12,dd3ff191-1acd-4c78-a96f-a026c870d6bc,2017-05-01,545,other,1,1,,0,0,0,0,0,0,0
 a12,eccae16c-866c-43d8-8021-c5d9238794c7,2017-05-01,257,lactating,1,1,,16,1,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,1,1,0,1,,0,0,1,1,0,2017-03-01,0,obc,no,no,yes,,,,,,,,,,,,1,1,,,2017-08-08,,,,,,,,,,1,,,,,,,2017-04-12,,1996-01-08,0,ee80d394-995e-447c-912b-9e681e53f2e6
 a12,ed7e32da-5a16-4606-812a-ba30d208ba66,2017-05-01,281,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1994-01-08,0,ce57fb59-aa4c-4665-8d7b-8ff89e1ce8ad
-a12,f2600795-0236-45a7-ac6c-baff437681f4,2017-05-01,810,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1950-01-28,0,d65b5b3b-266c-4d5d-b571-494b6023a2e3
+a12,f2600795-0236-45a7-ac6c-baff437681f4,2017-05-01,809,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1950-01-28,0,d65b5b3b-266c-4d5d-b571-494b6023a2e3
```
There were some inconsistency in the age_in_month.
is age_in_month in '2017-04-01' is x then in ''2017-05-01' has to be x+1.
fixed in this commit

Rohit25negi commented on custom/icds_reports/tests/outputs/ccs_record_monthly_2017-04-01_sorted.csv:
```diff
@@ -205,7 +205,7 @@ a19,39ff586d-a832-47a2-a28d-a7d7d6813729,2017-04-01,264,other,0,1,,0,0,0,0,0,0,0
 a19,9138bf62-3e9c-4fa6-a40c-9500a4ebf3b6,2017-04-01,240,other,0,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2017-08-31,0,sc,no,no,yes,,,,,,,,,,,,,,,,2017-08-31,,,,,,,,,,0,,,,,,,2017-05-12,,1997-05-08,0,bd72919c-742e-4b60-82d5-2d77d0dfa406
 a19,9a43a53e-2bfd-444f-b5de-3771669fb6ba,2017-04-01,265,other,0,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,1,0,,0,0,0,0,0,2017-03-20,0,sc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-05-12,,1995-04-09,0,49094a8c-f7b5-4367-8d6f-9214e8c701a3
-a19,a866de7b-13bc-400f-a81c-a6f38c2f7e13,2017-04-01,289,other,0,1,,0,0,0,1,0,0,0,0,3,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,1,0,,0,0,0,0,0,2017-04-20,0,sc,no,no,yes,,,,,,,,,,,,,,,,2017-05-08,,,,,,,,,,0,,,,,,,2017-05-12,,1993-04-10,0,9b544f35-f64a-4abc-becc-8e915ff86f52
+a19,a866de7b-13bc-400f-a81c-a6f38c2f7e13,2017-04-01,289,other,0,1,,0,0,0,1,0,0,0,0,3,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,1,0,,0,0,0,0,0,2017-04-02,0,sc,no,no,yes,,,,,,,,,,,,,,,,2017-05-08,,,,,,,,,,0,,,,,,,2017-05-12,,1993-04-10,0,9b544f35-f64a-4abc-becc-8e915ff86f52
```
Just corrected the add date. added in previous commits

Rohit25negi commented on custom/icds_reports/tests/outputs/ccs_record_monthly_2017-05-01_sorted.csv:
```diff
@@ -125,7 +125,7 @@ a14,b69790bc-a2b0-4ecc-a245-8dad9eff2fd9,2017-05-01,434,other,1,1,,0,0,0,0,0,0,0
 a14,ff673a47-44bf-4d2b-a35c-e91addd69b15,2017-05-01,,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,obc,,no,,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,,0,de188fca-40d1-4000-b34f-2257acddafc7
 a15,394a7f4a-7fe6-416d-8fd3-0cccdf84184a,2017-05-01,362,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,other,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1987-04-13,0,9fbe6f92-8cf6-4a9a-98ad-f22c75ce461d
-a15,3d1bdefc-a217-455d-af18-260f39f698f0,2017-05-01,338,pregnant,0,1,3,21,1,0,0,0,0,0,0,,1,1,0,0,0,1,1,1,1,,0,1,0,0,0,0,0,0,0,1,0,0,0,0,,1,1,0,0,0,2017-06-02,0,other,no,no,yes,,,,,,,,,,,,1,1,,,2017-08-08,,,,,,,,,,1,,,,,,,2017-05-02,,1989-04-12,0,92e7f715-1953-4bd2-abdd-e995b54c8a01
+a15,3d1bdefc-a217-455d-af18-260f39f698f0,2017-05-01,338,pregnant,1,1,3,21,1,0,0,0,0,0,0,,1,1,0,0,0,1,1,1,1,,0,1,0,0,0,0,0,0,0,1,0,0,0,0,,1,1,0,0,0,2017-06-02,0,other,no,no,yes,,,,,,,,,,,,1,1,,,2017-08-08,,,,,,,,,,1,,,,,,,2017-05-02,,1989-04-12,0,92e7f715-1953-4bd2-abdd-e995b54c8a01
```
by definition a women is considered pregnant when open_in_month is 1.

Rohit25negi commented on custom/icds_reports/tests/outputs/ccs_record_monthly_2017-04-01_sorted.csv:
```diff
@@ -3,7 +3,7 @@ a1,0f2ecc69-9f03-46e1-95fc-f28df2d5951d,2017-04-01,280,other,1,1,,0,0,0,0,0,0,0,
 a1,32890569-edc5-43ea-94df-430945afef1d,2017-04-01,472,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,1,,0,0,0,0,0,2016-05-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1978-01-16,0,0ee12d13-6321-48e0-86d7-426cc7c58b96
 a1,346ad0e7-210c-4b6f-9659-534f473922d1,2017-04-01,340,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,,1989-01-10,0,d1bd8a2f-f874-4a08-aeaf-41a65ec7b5f1
-a1,387085d6-7427-48c3-aee3-ff12819def46,2017-04-01,316,lactating,1,1,,20,1,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,1,1,0,2017-03-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,1,,,,,,,2017-04-12,,1991-01-10,0,a9bc2bb8-0f4f-4e21-a51e-39369d83fb1c
+a1,387085d6-7427-48c3-aee3-ff12819def46,2017-04-01,316,lactating,1,1,,20,1,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,1,1,0,2017-03-01,0,obc,no,no,yes,,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,1,,,2017-03-04,,,,2017-04-12,2017-03-04,1991-01-10,0,a9bc2bb8-0f4f-4e21-a51e-39369d83fb1c
```
bp_date  and home_visit_date contains the same info. Not sure why home_visit_date was added.
Though just adding those date info

calellowitz replied:
A lot of these dates are in the future (after the month of this table) that should not be possible

Rohit25negi replied:
you mean bp_date and  home_visit_date? @calellowitz .

Rohit25negi replied:
if you are saying about add and edd. Its possible to a to have edd in future because its expected delivery and add can be because then only definition of pregnant exisits which  is like 'if valid_in_month and add = '' or add > end of iteration month then 1 else 0'

Rohit25negi replied:
confirmed about edd. else definition of trimester cant exist

Rohit25negi replied:
reduced the edd for some of the rows. to make it closed to add to make it logically correct. But still I am pretty sure edd can be in future

calellowitz replied:
There are two dates in this file that appear in the future some rows. I cant really tell what they are, but they are the ones here shown as `2017-08-08` and `2017-03-04` (second and third added dates). The first is is the same for almost every case in the table, the latter is in the past for this case but the future (`2017-05-12`) for the case 4 down from this one. You are correct that EDD can be in the future, not sure what the other date is but i am surprised there are so many in the future.

Rohit25negi replied:
first one is which is `2017-08-08` is add and second one `2017-03-04` is  edd as discussed edd can be in future.
about the other date i.e `add`. I think technically 'add' can be in future for aggregation but not necessarily in the agg tables because suppose a case is opened in Aug-2018 and now is Jan-2019. She was pregnant and in Jan-2019 she delivered. So add will be say 2019-01-09. Now, when on Jan-10 2019 aggregation will run for Dec-2018 `add` would be already in future for Dec aggregation but in Dec She should be pregnant so this definition of `pregnant` handles [this](https://docs.google.com/spreadsheets/d/1RDLNIHF3iRkiQtWyrQjKH6f1bStF_S_9woXeITiMJKg/edit#gid=551271206&range=A26) condition( already in the coded in this PR)

Now to  exclude the future `add` date values from the agg table  we can set `add` only when `add` is < end of iteration month similar to date_death field here.
Does it sound reasonable?

Rohit25negi commented on custom/icds_reports/tests/outputs/ccs_record_monthly-st1-2017-05-01.csv:
```diff
@@ -7,7 +7,7 @@ a1,387085d6-7427-48c3-aee3-ff12819def46,2017-05-01,317,lactating,1,1,,15,1,0,0,0
 a1,632c7cb4-1612-4b4a-8ca9-95d910706bfd,2017-05-01,305,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,sc,no,no,yes,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,0,1992-01-10,,st1,d1,b1,a1,a1
 a1,75be0b2a-49dc-4bb5-b1c3-d9573e57dc0d,2017-05-01,352,other,1,1,,0,0,0,0,0,0,0,0,,0,0,0,0,0,1,0,0,0,,0,0,0,0,0,0,0,0,0,0,0,0,0,0,,0,0,0,0,0,2016-05-01,0,st,no,no,yes,,,,,,,,,,,,,,,2017-08-08,,,,,,,,,,0,,,,,,,2017-04-12,0,1988-02-11,,st1,d1,b1,a1,a1
-a1,7772c727-6482-4cd4-a81e-e42c759a00f7,2017-05-01,281,pregnant,1,1,2,15,1,0,0,0,0,0,0,,1,1,0,0,0,1,1,1,1,,1,0,0,0,0,0,0,0,1,0,0,0,0,0,,1,1,0,0,0,2017-08-31,0,sc,no,no,yes,,,,,,,,,,,,,,,2017-08-31,,,,,,,,,,1,,,2017-05-04,,,,2017-05-12,0,1994-01-08,,st1,d1,b1,a1,a1
+a1,7772c727-6482-4cd4-a81e-e42c759a00f7,2017-05-01,281,pregnant,1,1,2,15,1,0,0,0,0,0,0,,1,1,0,0,0,1,1,1,1,,1,0,0,0,0,0,0,0,1,0,0,0,0,0,,1,1,0,0,0,2017-08-31,0,sc,no,no,yes,,,,,,,,,,,,1,1,,2017-08-31,,,,,,,,,,1,,,2017-05-04,,,,2017-05-12,0,1994-01-08,,st1,d1,b1,a1,a1
```
this is same  the change done in previous commits

Rohit25negi commented on custom/icds_reports/tests/reports/test_awc_reports.py:
```diff
@@ -2499,10 +2499,12 @@ def test_awc_report_pregnant_first_record(self):
             self.assertEqual(
                 len(data['data']),
-                1
+                2
```
This is set to '2'. because from where data is being fetched there `.exclude(open_in_month==False)`
is used. And But definition is a case is not open in month it cannot be pregnant. And Earlier pregnant cases were 2 but because of exclusion above one was not coming. Since I corrected the conflict in test data hence this is the result of that.

calellowitz commented on custom/icds_reports/tests/reports/test_awc_reports.py:
```diff
@@ -2550,11 +2552,11 @@ def test_pregnant_details_first_record_second_trimester(self):
-                        'edd': 'Data Not Entered',
-                        'home_visit_date': 'Data Not Entered',
+                        'edd': datetime.date(2017, 8, 31),
+                        'home_visit_date': datetime.date(2017, 5, 04),
```
@Rohit25negi The 4 needs be `4` not `04`

Rohit25negi replied:
resolved already

Rohit25negi committed:
Make edd close to add in past
To make it logically correct

calellowitz commented on custom/icds_reports/utils/aggregation_helpers/ccs_record_monthly.py:
```diff
@@ -52,62 +52,139 @@ def tablename(self):
+                                                                                       start_month_string)
+        alive_in_month = "(case_list.date_death is null OR case_list.date_death-{}>0)".format(
+            start_month_string)
+        seeking_services = "(case_list.is_availing = 1 AND case_list.is_migrated = 0)"
```
does it not need to be open to be seeking services? does the old version allow this?

Rohit25negi replied:
yeah I think So if I consider the definition of seeking in child_health and ccs_Record is same

calellowitz commented on custom/icds_reports/utils/aggregation_helpers/ccs_record_monthly.py:
```diff
@@ -52,62 +52,139 @@ def tablename(self):
-            ('extra_meal', 'ucr.extra_meal'),
-            ('resting_during_pregnancy', 'ucr.resting_during_pregnancy'),
-            ('bp_visited_in_month', 'ucr.bp_visited_in_month'),
+            ('thr_eligible', 'CASE WHEN {} THEN 1 ELSE 0 END'.format(valid_in_month)),
```
you only have to be valid in month to be thr eligible? why was it a separate ucr field before?

Rohit25negi replied:
so valid in month is made up of is (pregnant or lactating) which further are broken down to more  parameters . here: https://github.com/dimagi/commcare-hq/pull/22752/files/6de226f5336b42ca7d4a5a32741dcd07179f1cf4#diff-3fdee4e410a43f289000bcdf4802c62aR82.

The reason for It to be  a separate UCR field before is not clear to me. It seems redundant to me to store it as UCR field. One Reason I could think of is that earlier no calculation was done outside of UCR.

calellowitz commented on custom/icds_reports/utils/aggregation_helpers/birth_preparedness_forms.py:
```diff
@@ -71,31 +78,40 @@ def aggregation_query(self):
+            GREATEST(ucr.play_family_planning_vid, prev_month.play_family_planning_vid) as play_family_planning_vid,
+            GREATEST(ucr.conceive,prev_month.conceive) as conceive,
+            GREATEST(ucr.counsel_accessible_ppfp, prev_month.counsel_accessible_ppfp) as counsel_accessible_ppfp,
+            COALESCE(ucr.ifa_last_seven_days, prev_month.ifa_last_seven_days) as ifa_last_seven_days,
```
using the previous month's `ifa_last_seven_days` seems wrong because it is definitely not the last 7 days. is that what we are doing wrong?

Rohit25negi replied:
hey I did this acc to the definition. Which says latest BP form submitted before the end of month. does not it makes sense to include the previous month ?
https://docs.google.com/spreadsheets/d/1RDLNIHF3iRkiQtWyrQjKH6f1bStF_S_9woXeITiMJKg/edit#gid=551271206&range=B40

calellowitz commented on custom/icds_reports/utils/aggregation_helpers/ccs_record_monthly.py:
```diff
@@ -52,62 +52,139 @@ def tablename(self):
+                                 "THEN 3 ELSE 2 END END)"
+
+        postnatal = "(case_list.add is not null AND ({}-case_list.add)::integer>=0 AND " \
+                    "({}-case_list.add)::integer<=21)".format(end_month_string, start_month_string)
```
what is this indicator?

Rohit25negi replied:
hey this is mother till 21 days after delivery
https://docs.google.com/spreadsheets/d/1RDLNIHF3iRkiQtWyrQjKH6f1bStF_S_9woXeITiMJKg/edit#gid=551271206&range=B62

calellowitz commented on custom/icds_reports/utils/aggregation_helpers/ccs_record_monthly.py:
```diff
@@ -157,25 +234,29 @@ def aggregation_query(self):
             ('home_visit_date', 'agg_bp.latest_time_end_processed'),
-            ('date_death', 'person.date_death'),
-            ('person_case_id', 'ucr.person_case_id')
+            ('date_death', 'person_cases.date_death'),
```
I think you only want to add this if date_death is before month end

Rohit25negi replied:
Yeah correct

calellowitz commented on custom/icds_reports/utils/aggregation_helpers/ccs_record_monthly.py:
```diff
@@ -52,62 +52,139 @@ def tablename(self):
+            ('lactating_all', 'CASE WHEN {} THEN 1 ELSE 0 END'.format(lactating_all)),
+            ('institutional_delivery_in_month', 'CASE WHEN agg_delivery.where_born=3 AND {} THEN'
+                                                ' 1 ELSE 0 END'.format(delivered_in_month)),
+            ('add', 'case_list.add'),
```
I think you only want to add this if `add` is beofre month end

Rohit25negi replied:
`delivered_in_month` ensured that.

Rohit25negi replied:
sorry, I somehow saw this comment was applied on 180 line. resolving this

calellowitz commented on custom/icds_reports/utils/aggregation_helpers/ccs_record_monthly.py:
```diff
@@ -157,25 +234,29 @@ def aggregation_query(self):
-            LEFT OUTER JOIN "{person_cases_ucr}" person ON person.doc_id = ucr.person_case_id
-            WHERE ucr.month = %(start_date)s
-            ORDER BY ucr.awc_id, ucr.case_id
+            FROM "{ccs_record_case_ucr}" case_list
```
I think you want to add filter to only include people opened before the end of this month. Otherwise there will be rows from future cases

Rohit25negi replied:
Sure,I can do that but that will lead to deletion of all the rows from output data which have open_data > end of iteration month. Does that sound right

Rohit25negi replied:
yes there does exists data from future in `ccs_record_monthly` table. Running this query 
`select pregnant,valid_in_month,opened_on from "ccs_record_monthly_2018-11-01" where add>'2018-11-30' limit 1`
Ideally should return no rows. But it has the records opened in month `Dec` even though this is the table for `Nov`.

So, If we think this should not be the case from this change onward. I can correct it by adding filters. But that would need deletion of future rows from output test data
@calellowitz

calellowitz reviewed (changes requested):
Few changes but generally speaking looks much better than before. How heavily tested is this?

Rohit25negi commented on custom/icds_reports/tests/reports/test_awc_reports.py:
```diff
@@ -2499,10 +2499,12 @@ def test_awc_report_pregnant_first_record(self):
             self.assertEqual(
                 len(data['data']),
-                1
+                2
```
reason: https://github.com/dimagi/commcare-hq/pull/22752/files/3aa79112bcdca4721e1c18672041b928766654cd..d4458f662ba021a6b162e831b9a3cabda8eef871#r247127088

Rohit25negi commented on custom/icds_reports/ucr/data_sources/ccs_record_cases.json:
```diff
@@ -521,6 +545,15 @@
         "column_id": "open_person_count"
       },
+      {
+        "column_id": "pregnant",
```
yeah right. just forgot to remove it from UCR.

Rohit25negi commented:
> Few changes but generally speaking looks much better than before. How heavily tested is this?

About test scope. I tried to include all the existing test conditions being produced using the fixture. which is I guess quite good.  Also needed, little corrections in fixtures here and there.

emord commented on custom/icds_reports/ucr/data_sources/ccs_record_cases.json:
```diff
@@ -503,6 +503,30 @@
         }
       },
+      {
+        "column_id": "person_case_id",
```
The UCR changes will need to be deployed before these aggregation changes, so that when the aggregation runs the first time, these columns aren't empty. You will also need to write a script to migrate data into this as a naive rebuild may take too long. You should be able to grab the information from the ccs record monthly table. For example:

```sql
UPDATE ccs_record_cases_ucr  ccs_ucr SET
   person_case_id = monthly.person_case_id
FROM ccs_record_monthly_tableau AS monthly
WHERE ccs_ucr.person_case_id IS NULL AND ccs_ucr.doc_id = monthly.doc_id
```

That's the general idea at least, there may be a syntax error there, and you'll want to update all the properties instead of just one.

Rohit25negi replied:
hmmmm. Thanks Emord, in that case will separate this ucr update to other PR.

Rohit25negi committed:
Merged with master

Rohit25negi committed:
Added the exptected data
filter was applied on home_visit_date to include only when home_visit_date is set

Rohit25negi committed:
remove pregnant field from UCR

Rohit25negi commented:
its up for review. but not to merge before https://github.com/dimagi/commcare-hq/pull/22973/files

Rohit25negi committed:
Removed Pregnant column from fixture

calellowitz reviewed (changes requested):
It looks like most of my questions from last time were not addressed. So this is mostly just the same review. I dont think i caught anything additional

Rohit25negi committed:
Merged with master

Rohit25negi commented on custom/icds_reports/tests/reports/test_awc_reports.py:
```diff
@@ -2572,7 +2594,7 @@ def test_awc_report_lactating_first_record(self):
                     'age': 20,
                     'num_pnc_visits': None,
-                    'add': None,
+                    'add': datetime.date(2017, 3, 1),
```
It came because `add`  is added to the fixture which was not present in there before.

Rohit25negi commented on custom/icds_reports/tests/reports/test_awc_reports.py:
```diff
@@ -2540,7 +2542,27 @@ def test_pregnant_details_first_record_second_trimester(self):
             )
             self.assertEqual(
                 data['data'][1],
-                []
```
This has changed Because Lukaz added a filter to consider only the data which has home_visit_date(which is same as bp_date) set and It was not being set earlier because fixture was not there.  I added the row in the fixture to pass the test cases. Therefore now home_visit_date is set. Hence, this change

Rohit25negi replied:
not sure what you meant by `It looks liek the same case is loaded twice`

Rohit25negi committed:
changes dependency of migration file

Rohit25negi commented:
hey @calellowitz  based on our last offline discussion on your comments. Please review  this

Rohit25negi committed:
Merged with master

emord commented on custom/icds_reports/models/aggregate.py:
```diff
@@ -1443,6 +1470,12 @@ class AggregateCcsRecordDeliveryForms(models.Model):
+    where_born = models.PositiveSmallIntegerField(
+        null=True,
+        help_text="Where the child is born",
+        default=0
```
I think this will be changed in the other PR?

emord commented on custom/icds_reports/utils/aggregation_helpers/ccs_record_monthly.py:
```diff
@@ -157,26 +234,29 @@ def aggregation_query(self):
-            WHERE ucr.month = %(start_date)s
-            ORDER BY ucr.awc_id, ucr.case_id
+            FROM "{ccs_record_case_ucr}" case_list
+            LEFT OUTER JOIN "{person_cases_ucr}" person_cases ON case_list.person_case_id = person_cases.doc_id
```
Was there a particular reason to change the ordering of these joins?

Rohit25negi committed:
After merge with master

Rohit25negi committed:
check for not null

Rohit25negi committed:
Updated the test cases

Rohit25negi commented:
> @Rohit25negi The additional person cases will conflict with the change I made in #23198
> 
> I think that redoing your additions on that branch should be pretty easy as each row does not have many columns filled out, but feel free to ping me if you run into any issues.

Hey I saw that, I resolved that when merged with master
